[Xu Daofu] iP

therealdaofu opened the pull request:
### DukePro

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind from having to remember things you need to do. It's,

- text-based
- easy to learn
- SUPER FAST to use

All you have to do is:

1. download it from [here](https://github.com/therealdaofu/ip)
2. run the jar file
3. add your tasks
4. let it manage your task for you :)

_Best part_? It is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the main method:
```
    public static void main(String[] args) {
        new Duke("C:\\Users\\xudao\\OneDrive\\Documents\\NUS FILES\\" +
                "year 2\\sem 1\\cs2103t\\git\\ip\\src\\data\\storage.txt").run();
    }
```

Commit:
Add Gradle support

therealdaofu committed:
Level-1

therealdaofu committed:
Level-2

therealdaofu committed:
Level-3

therealdaofu committed:
level-3.1 improved asethetic of bot, added error message for out of bound index input

therealdaofu committed:
level-4

therealdaofu committed:
A-TextUiTesting

therealdaofu committed:
level-5

therealdaofu committed:
level-6

therealdaofu committed:
level-6.1

therealdaofu committed:
level-6-final

therealdaofu committed:
Update Memory class

Memory class stores tasks in a ArrayList, this is reset every time
the chatbot is ran, does not allow user to store their tasks
for later.

Tasks are now also stored in an external file to allow access
across different runs of the chatbot.

therealdaofu committed:
Add javadoc comments

therealdaofu committed:
Update Deadline class

deadline class took in and stored date and time as strings

now it will take in date and time in the ISO-8601 format

therealdaofu committed:
Update Deadline Class to store date and time as java.time.LocalDate

Deadline class used to take in and store date as String

Now it takes in date and time in the ISO-8601 format

therealdaofu committed:
Revert "Update Deadline class"

This reverts commit 7f9e696e06ef758042537abef15d067b53469262.

therealdaofu committed:
Merge remote-tracking branch 'origin/level-7'

therealdaofu committed:
Merge remote-tracking branch 'origin/b8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java

therealdaofu committed:
Fix bug in Memory class

therealdaofu committed:
Split Memory class into Storage and TaskList class, add Ui class

Memory class was in charge of both manupilating the stored tasks as well as saving the tasks to external file.

Now storage class handles the saving and reading of data from the external file.
TaskLists handles the manipulation, including adding, removing and marking etc.
Ui handles all interaction with user.

therealdaofu committed:
Add all classes to myDuke package

therealdaofu committed:
Add task package

therealdaofu committed:
Rename package to myduke

therealdaofu committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

therealdaofu committed:
Fix bug in Ui.response: exception should be thrown properly
now when index given is out of bound

therealdaofu committed:
Add JUnit Tests
Add tests for deadline, event and todo class
Not done: add tests for storage, tasklist, ui class

therealdaofu committed:
Package App as JAR file

therealdaofu committed:
Add specific exceptions, move all exceptions to exception package

therealdaofu committed:
Add JavaDoc comments to all classes

therealdaofu committed:
Tweak code to comply with coding standard

therealdaofu committed:
Add Ui.find method to allow user to find a task by searching with
a keyword.

Add MissingKeyWordException for the case when user does not give
a keyword

therealdaofu committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Tweak code to comply with coding standard

therealdaofu committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add Ui.find method to allow user to find a task by searching with a keyword.

# Conflicts:
#	src/main/java/myduke/Storage.java
#	src/main/java/myduke/TaskList.java

therealdaofu committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/exception/DukeException.java
#	src/main/java/myduke/Duke.java
#	src/main/java/myduke/Storage.java
#	src/main/java/myduke/TaskList.java
#	src/main/java/myduke/Ui.java
#	src/main/java/task/Deadline.java
#	src/main/java/task/Event.java
#	src/main/java/task/Task.java
#	src/main/java/task/ToDo.java

therealdaofu committed:
Tweaked code to follow coding standard

therealdaofu committed:
Edit code to comply with coding style

zicotjia commented on src/main/java/exception/MarkException.java:
```diff
@@ -0,0 +1,14 @@
+/**
+ * This exception is thrown when the user is trying to mark a task that has already been marked.
+ */
+public class MarkException extends DukeException {
```
For better clarity, maybe prefix all Exception extending DukeException with "Duke". i.e DukeMarkException.

zicotjia commented on src/main/java/myduke/Duke.java:
```diff
@@ -0,0 +1,66 @@
+    public static void main(String[] args) {
+        //String filepath = "C:\\Users\\xudao\\OneDrive\\Documents\\NUS FILES\\" +
+        //        "year 2\\sem 1\\cs2103t\\git\\ip\\src\\data\\storage.txt"
+        String filepath = "C:\\Users\\xudao\\Documents\\nus\\year 2\\sem 1\\2103t"
+                + "\\git stuff\\individual project\\src\\data\\storage.txt";
```
Perhaps you can use relative path instead of absolute path or to store the path in a .env file. This is so that Duke can find the filePath in the other computer

zicotjia commented on src/main/java/myduke/Ui.java:
```diff
@@ -0,0 +1,207 @@
+     * @param taskLists TaskList used to store the tasks.
+     * @param storage   Storage used to save to file.
+     */
+    public Ui(TaskList taskLists, Storage storage) {
```
Tasklist should be singular.

zicotjia commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,45 @@
+     */
+    @Override
+    public String getDescription() {
+        String status = super.getStatus() ? "T" : "F";
```
I like that you use ternary operator for simple cases.

zicotjia reviewed (commented):
The overall variable naming is very good

wadia82 commented on src/main/java/exception/OutOfBoundIndexException.java:
```diff
@@ -0,0 +1,14 @@
+/**
+ * This exception is thrown when user is trying to access a task with an out of bound index.
+ */
+public class OutOfBoundIndexException extends DukeException {
```
You could start the class name with "Duke...." as in you could have a DukeOutOfBoundIndexException. This could be done for all other classes extending DukeException.

wadia82 commented on src/main/java/myduke/Duke.java:
```diff
@@ -0,0 +1,66 @@
+    public Duke(String filePath) {
+        isEnd = false;
+        Storage storage = new Storage(filePath);
+        TaskList tasks = new TaskList();
```
Object should be named 'task' instead of 'tasks'.

wadia82 commented on src/main/java/myduke/TaskList.java:
```diff
@@ -0,0 +1,174 @@
+ */
+public class TaskList {
+    //this is where the tasks are stored
+    private final ArrayList<Task> taskLists;
```
You could name the ArrayList tasks instead of taskLists as it stores tasks.

wadia82 commented on src/main/java/task/Event.java:
```diff
@@ -0,0 +1,41 @@
+ * This class encapsulates the idea of a event.
+ */
+public class Event extends Task {
+    private final String duration;
```
Variable name should be in uppercase since it is final.

wadia82 commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,70 @@
+ * This class encapsulates the idea of a generic task.
+ */
+public class Task {
+    private final String description;
```
Variable name should be in uppercase since it is final.

wadia82 reviewed (commented):
Overall, I liked the structure of your program. The OOP was well constructed. There are some minor coding standard violations but they are mostly with variable naming.

therealdaofu committed:
no message

therealdaofu committed:
Add GUI to Duke
Remove text-based UI
Compact DukeExceptions back into 1 class

therealdaofu committed:
Add parser class and Command classes

Ui class handled parsing user input, deciding what to do with
user input, and also returning String messages to inform the user.

Parsing of user input and decision marking has been extracted and
placed in their own classes for better control.

Parser class takes in user input and returns the appropriate
command object

All command classes extends from abstract class Command,
which allows polymorphism when dealing with the command returned
by Parser.

therealdaofu committed:
Add assert statements

therealdaofu committed:
Improve code quality

Abstracted out methods from parse method in Parser class.
Previously, parse method was too long and had weak SLAP,
processing the user input has now been abstracted out into multiple
other methods to maintain good SLAP.

Made the happy path prominent in Tasklist.
Previously, methods in tasklist class handle the main path first,
and checked for unusual cases later.
Now, methods check for unusual conditions immediately and keeps
the main path un-indented

Edit default branch in loadfromfile method in storage
to throw an error.

therealdaofu committed:
Merge pull request #1 from therealdaofu/branch-A-Assertions

Add assert statements

therealdaofu committed:
Merge branch 'master' of https://github.com/therealdaofu/ip

therealdaofu committed:
Merge pull request #2 from therealdaofu/branch-A-CodeQuality

Improve code quality

therealdaofu committed:
Merge branch 'master' of https://github.com/therealdaofu/ip

therealdaofu committed:
Add Help command
provides user with a list of valid commands and formats

therealdaofu committed:
Add more personality to chat bot.
Chat bot renamed to Koba
Edit JavaDoc comments for most classes.

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
trim input

therealdaofu committed:
Merge remote-tracking branch 'origin/master'

therealdaofu committed:
Add error handling when save file is missing
Storage now able to create new save file if save is missing.

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Improve UI
app is able to be resized, textbox auto expands based on message
length.

therealdaofu committed:
Merge branch 'master' of https://github.com/therealdaofu/ip

therealdaofu committed:
no message

therealdaofu committed:
no message

therealdaofu committed:
no message

therealdaofu committed:
Update README.md

therealdaofu committed:
Fix bug with find command

therealdaofu committed:
Merge branch 'master' of https://github.com/therealdaofu/ip

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Update README.md

therealdaofu committed:
Merge branch 'master' of https://github.com/therealdaofu/ip

therealdaofu committed:
Fix save file not being created properly
koba.jar now able to create a folder data and a text file in
it to store the data.

Fix bugs for help, list and bye commands
Previously, typing these commands followed by random letters will
still run the command.
Now, an error will be shown to the user

therealdaofu committed:
fix table

therealdaofu committed:
Update README.md
